dropbox/dropbox.py

```diff
@@ -452,6 +466,39 @@ def request_json_string(self,
         else:
             raise HttpError(request_id, r.status_code, r.text)
 
+    def with_path_root(self, path_root=None):
+        """
+        Creates a new object with the same parameters as this object, but also
+        sets the ``Dropbox-API-Path-Root` header to :param:`path_root`.
+
+        :param PathRoot path_root: the path root object to serialize and use for
+        the ``Dropbox-API-Path-Root`` header. (If ``None``, the header is
+        cleared.)
+        :return: a new instance of this class with the ``Dropbox-API-Path-Root``
+        header set.
+        """
+        headers = self._headers
+
+        if headers is not None:
+            headers = dict(headers)  # make a copy that we can manipulate
+            headers.pop(PATH_ROOT_HEADER)
+
+        if path_root is not None:
+            if headers is None:
+                headers = {}
+            headers[PATH_ROOT_HEADER] = stone_serializers.json_encode(PathRoot_validator, path_root)
+
+        # FRAGILE: This depends on subclasses not adding arguments to their own __init__ methods
+        return type(self)(
+            oauth2_access_token=self._oauth2_access_token,
+            max_retries_on_error=self._max_retries_on_error,
+            max_retries_on_rate_limit=self._max_retries_on_rate_limit,
+            user_agent=self._user_agent,
+            session=self._session,
+            headers=headers,
+            timeout=self._timeout,
+        )
```

Comment:
Why return a duplicate instance of the `dbx` object instead of modifying the headers of the existing object?

Reply:
@qimingyuan can probably speak more intelligently on this topic. The TL;DR is that these object are largely treated as lightweight and immutable. This approach keeps with that convention.

Reply:
There _are_ cases where one would want to modify a path root header for an existing transport cases in response to an error received from the server, so I _think_ having this method makes sense. That being said, I've added two additional methods (`set_path_root_header` and `clear_path_root_header`) to assist with constructing new objects.